handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java

```diff
@@ -1056,11 +1054,26 @@ public synchronized void beginHandshake() throws SSLException {
                 // for renegotiation.
 
                 handshakeState = HandshakeState.STARTED_EXPLICITLY; // Next time this method is invoked by the user,
-                                                          // we should raise an exception.
+                                                                    // we should raise an exception.
                 break;
-            case FINISHED:
             case STARTED_EXPLICITLY:
-                throw RENEGOTIATION_UNSUPPORTED;
+                // Nothing to do as the handshake is not done yet.
+                break;
+            case FINISHED:
+                if (clientMode) {
+                    // Only supported for server mode at the moment.
+                    throw RENEGOTIATION_UNSUPPORTED;
+                }
+                if (SSL.renegotiate(ssl) != 1 || SSL.doHandshake(ssl) != 1) {
```

Comment:
thanks ... good to capture your "research" into openssl so we don't have to wonder why this was done in 1 months time :)